Add SettingsRepo, refactoring & setting for battery limit

miguelguzmanr opened the pull request:
Things to note:
- Still testing these changes. My sincerest apologies if anything broke.
- Could not decide on a name or correct folder for the SettingsRepo class. Please change.
- Could not decide on a proper name for functions that use "cat"/"echo" commands in the ShellExecutor class to read and write settings. Not sure if they're even useful at all, either, at least currently not very much. Please change.

langerhans commented on app/src/main/java/de/langerhans/odintools/tools/DeviceUtils.kt:
```diff
@@ -4,20 +4,20 @@ import de.langerhans.odintools.tools.DeviceType.*
 
+    fun getDeviceVersion() = executor.getStringProperty(SettingsRepo.KEY_BUILD_VERSION, "")
+
+    fun getDeviceVendor() = executor.getStringProperty(SettingsRepo.KEY_VENDOR_NAME, "")
```
Technically this value is not the vendor. Suggestion: `getDeviceCodename()`

miguelguzmanr replied:
Sounds good!

langerhans commented on app/src/main/java/de/langerhans/odintools/tools/SettingsRepo.kt:
```diff
@@ -0,0 +1,110 @@
+) {
+
+    init {
+        println("Settings init")
```
For logging use `Log.d()` or other levels. Maybe we should think about adding a logging library like Timber to simplify this.

miguelguzmanr replied:
Oops! Slipped from my sight. Good catch!

langerhans commented on app/src/main/java/de/langerhans/odintools/tools/SettingsRepo.kt:
```diff
@@ -0,0 +1,110 @@
+    private val executor: ShellExecutor,
+) {
+
+    init {
```
I'd make this call explicitly somewhere where it makes most sense, e.g. OdinToolsApplication. A place that's only run once at startup. Right now this would be called for every class that injects an instance of SettingsRepo right?

miguelguzmanr replied:
Actually, this is what the `println('Settings init') was meant for. Wanted to confirm if SettingsRepo was being initialized for every class that injects an instance. Not sure if this is the best way to confirm this, but can agree that making this call explicitly would make more sense.

langerhans commented on app/src/main/java/de/langerhans/odintools/tools/SettingsRepo.kt:
```diff
@@ -0,0 +1,110 @@
+
+    private fun enableA11yService() {
+        val currentServices =
+            executor.executeAsRoot("settings get secure enabled_accessibility_services")
```
Use a KEY_ constant for this key as well?

miguelguzmanr replied:
Sounds good!

langerhans commented on app/src/main/java/de/langerhans/odintools/tools/SettingsRepo.kt:
```diff
@@ -0,0 +1,110 @@
+        get() = executor.getStringProperty(KEY_VENDOR_NAME, "")
+
+    var saturation: Float
+        get() = throw UnsupportedOperationException()
```
I'm a bit torn on this. Style wise a property that can't even be read shouldn't be a property at all. This would only fail at runtime, not at compile time so I'd rather use a method for this setting tbh.

miguelguzmanr replied:
Sounds good. Just to clarify, the current saturation value will be recovered from the user-defined setting (shared preferences).

langerhans replied:
Yup that works. I also just looked into SurfaceFlinger and it reads `persist.sys.sf.color_saturation` on boot to apply the setting. So in theory we could write that and get rid off the persistence in OdinTools. Then again I like the ide of it not applying anymore when someone decides to uninstall OdinTools. Leaves the device in a clean state. 🤔

langerhans reviewed (changes requested):
For the name I think it's fine as it is. I'm bad at naming things so I don't worry so much about it. As for the package maybe we can move SettingsRepo and ShellExecutor to a new package "system" since they are responsible for communicating with the system. Keep DeviceUtils in tools and BootReceiver can live there for now as well. Maybe that will be a bit cleaner.
Otherwise nice work, thanks!

miguelguzmanr commented:
Introduces configurable charging separation.
Requires thorough testing, as this may introduce additional battery drain due to having to monitor the battery level.

![image](https://github.com/langerhans/OdinTools/assets/40463442/48c03dec-67a5-4200-ab97-6ff36562999e)

miguelguzmanr commented:
This approach only works if the app is running in the background. I think this might need to be implemented as a foreground service or a background task.

langerhans commented:
Even with the fix from #23 this does not work in background? Could we just watch battery state with `Intent.ACTION_BATTERY_CHANGED`?

isedwards commented:
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
d4rk4444, @sarog, @Samuel2314, @ManKiam, @snripa, @unchang3d, @sdmg2024, @kir0618, @OfficialChaos, @Dussed, @brenopolanski, @AndyChenW, @Mauker1, @TimDaub, @Interceptorlx, @sqqihao, @bitcoinprecept, @bgdshka, @lapis885, @Osukanmi, @kpallen1, @azxcv123vc456, @sxb1n9, @hypomn, @apzentral, @ieee820, @Bowen-Down, @felipefrizzo, @tuxcanfly, @deloxd, @eren12331, @retaw, @DiabloDaN, @whitebaby, @chaeplin, @nsfestone, @timmygee, @zeta009, @Mita234, @ajnadox, @Vertikar, @zacke, @ImShalev, @rodgermoore, @jeffhuys, @raymonddurk, @haofree, @blkrt, @haromz, @xxlv

miguelguzmanr commented:
Introduces a service to monitor the battery level.
Just trying out a different approach.
Sadly, the service forcefully ends when the app is terminated in any way.

Requires #23 to work in the background

![image](https://github.com/langerhans/OdinTools/assets/40463442/93711dbe-13b1-49e8-814b-517c82decfc8)

Found out that charging separation briefly turns off when the screen is locked. Not sure about the culprit. Added a trigger for screen off event to combat this behavior.
Should work properly now.

Added an option to dump a log file for debug builds.
Might be useful for troubleshooting issues.

![image](https://github.com/langerhans/OdinTools/assets/40463442/4d362685-836a-4f5d-8de2-0347ff300a4e)

miguelguzmanr commented:
@langerhans please let me know your thoughts!

langerhans commented on app/src/main/java/de/langerhans/odintools/main/MainActivity.kt:
```diff
@@ -205,6 +227,25 @@ fun SettingsScreen(viewModel: MainViewModel = hiltViewModel(), navigateToOverrid
+                    viewModel.updateChargeLimitPreference(it)
+                }
+            }
+            if (BuildConfig.DEBUG) {
```
Usually people wouldn't run debug builds no?

miguelguzmanr replied:
This is something I've been debating with myself a lot. I think I was looking for a reason to add this feature, which serves a very niche purpose on its own. As far as it goes, people usually wouldn't run debug builds. Please let me know if this is something that isn't very useful in its current state.

langerhans replied:
I didn't mean to say it's not useful cause it certainly is for cases where logs can help. Was just wondering if we should expose it for release builds as well :)

miguelguzmanr replied:
Resolved in 884c8dc

langerhans commented on app/src/main/java/de/langerhans/odintools/service/BatteryLevelMonitorService.kt:
```diff
@@ -0,0 +1,97 @@
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class BatteryLevelMonitorService @Inject constructor() : Service() {
```
This might be a dumb question but can our accessibility service do the same work? Would spare us from having to run a foreground service.

miguelguzmanr replied:
Great question! I haven't tested this yet. Will give it a try and report back.

miguelguzmanr replied:
Resolved in 8d99eea

langerhans reviewed (approved):
Happy to merge this, but just had two more questions :)

miguelguzmanr commented on app/src/main/java/de/langerhans/odintools/main/MainActivity.kt:
```diff
@@ -26,16 +26,28 @@
 import androidx.navigation.compose.composable
 import androidx.navigation.compose.rememberNavController
 import dagger.hilt.android.AndroidEntryPoint
+import de.langerhans.odintools.BuildConfig
```
Fixed in 8d99eea

miguelguzmanr commented:
@langerhans please let me know if this approach works better!

langerhans commented:
Yup looks good to me, LGTM
